File: web/server/utils/logger.ts
/** 管理 Canvas MCP 的终端 Debug 日志。 */
class Logger {
    readonly enabled = process.argv.includes("--debug");

    /** 输出 Debug 级别日志，仅在 --debug 模式生效。 */
    debug(message: string, details?: unknown) {
        if (this.enabled) this.write("DEBUG", message, details);
    }

    /** 输出 Info 级别日志。 */
    info(message: string, details?: unknown) {
        this.write("INFO", message, details);
    }

    /** 输出 Warn 级别日志。 */
    warn(message: string, details?: unknown) {
        this.write("WARN", message, details);
    }

    /** 输出 Error 级别日志。 */
    error(message: string, details?: unknown) {
        this.write("ERROR", message, details);
    }

    private write(level: string, message: string, details?: unknown) {
        const suffix = details === undefined ? "" : ` ${JSON.stringify(sanitize(details))}`;
        console.log(`${new Date().toISOString()} ${level} ${message}${suffix}`);
    }
}

/** 清理日志内容中的敏感数据和 Data URL。 */
function sanitize(value: unknown, key = ""): unknown {
    if (/token|authorization|api.?key|dataurl/i.test(key)) return "[REDACTED]";
    if (typeof value === "string" && value.startsWith("data:")) return `[DATA URL ${value.length} chars]`;
    if (value instanceof Error) return { name: value.name, message: value.message };
    if (!value || typeof value !== "object") return value;
    if (Array.isArray(value)) return value.map((item) => sanitize(item, key));
    return Object.fromEntries(Object.entries(value as Record<string, unknown>).map(([field, item]) => [field, sanitize(item, field)]));
}

export const logger = new Logger();

File: mcp/src/canvas_mcp/logger.py
"""Debug-gated terminal logger mirroring ``web/server/utils/logger.ts``."""

from __future__ import annotations

import json
import os
import re
from datetime import UTC, datetime
from typing import Any

_SENSITIVE = re.compile(r"token|authorization|api.?key|dataurl", re.IGNORECASE)


def _sanitize(value: Any, key: str = "") -> Any:
    """Strip sensitive fields and inline data URLs from log details."""
    if key and _SENSITIVE.search(key):
        return "[REDACTED]"
    if isinstance(value, str) and value.startswith("data:"):
        return f"[DATA URL {len(value)} chars]"
    if isinstance(value, BaseException):
        return {"name": type(value).__name__, "message": str(value)}
    if isinstance(value, dict):
        return {str(field): _sanitize(item, str(field)) for field, item in value.items()}
    if isinstance(value, (list, tuple)):
        return [_sanitize(item, key) for item in value]
    return value


class Logger:
    """Timestamped logger whose ``debug`` output is gated by ``CANVAS_MCP_DEBUG``."""

    def __init__(self, enabled: bool = False) -> None:
        self.enabled = enabled

    def debug(self, message: str, details: Any = None) -> None:
        if self.enabled:
            self._write("DEBUG", message, details)

    def info(self, message: str, details: Any = None) -> None:
        self._write("INFO", message, details)

    def warn(self, message: str, details: Any = None) -> None:
        self._write("WARN", message, details)

    def error(self, message: str, details: Any = None) -> None:
        self._write("ERROR", message, details)

    def _write(self, level: str, message: str, details: Any = None) -> None:
        suffix = "" if details is None else f" {json.dumps(_sanitize(details), ensure_ascii=False, default=str)}"
        stamp = datetime.now(UTC).isoformat()
        print(f"{stamp} {level} {message}{suffix}", flush=True)


logger = Logger(os.environ.get("CANVAS_MCP_DEBUG", "").strip().lower() in {"1", "true", "yes", "on"})
